staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -315,23 +320,96 @@ func strategicPatchObject(
 
 type applyPatcher struct {
 	*patcher
+
+	model proto.Schema
 }
 
-func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
-	versionedObjToUpdate, err := p.creater.New(p.kind)
+// This is not going to stay, but we don't have a way to pass the
+// workflow id so far, so let's always use this one...
+const workflowId = "default"
+
+func (p *applyPatcher) LiveObject(obj runtime.Object) (map[string]interface{}, error) {
+	vo, err := p.unsafeConvertor.ConvertToVersion(obj, p.kind.GroupVersion())
 	if err != nil {
 		return nil, err
 	}
-	if err := runtime.DecodeInto(p.codec, p.patchBytes, versionedObjToUpdate); err != nil {
-		return nil, err
+	return runtime.DefaultUnstructuredConverter.ToUnstructured(vo)
+}
+
+func (p *applyPatcher) LastAppliedObject(obj runtime.Object, workflow string) (map[string]interface{}, error) {
+	acsor, err := meta.Accessor(obj)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get accessor: %v", err)
 	}
-	// For now, instead of performing a structured merge we will just attempt a PUT
-	// So we just convert the object back to unversioned (aka internal version).
-	unversionedObjToUpdate, err := p.toUnversioned(versionedObjToUpdate)
+	last := make(map[string]interface{})
+	if acsor.GetLastApplied()[workflow] != "" {
+		if err := json.Unmarshal([]byte(acsor.GetLastApplied()[workflow]), &last); err != nil {
+			return nil, fmt.Errorf("couldn't unmarshal last applied field: %v", err)
+		}
+	}
+	return last, nil
+}
+
+func (p *applyPatcher) PatchObject() (map[string]interface{}, error) {
+	patch := make(map[string]interface{})
+	if err := yaml.Unmarshal(p.patchBytes, &patch); err != nil {
+		return nil, fmt.Errorf("couldn't unmarshal patch object: %v (patch: %v)", err, string(p.patchBytes))
+	}
+	return patch, nil
+}
+
+func (p *applyPatcher) OutputObject(result interface{}) (runtime.Object, error) {
+	voutput, err := p.creater.New(p.kind)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to create empty output object: %v", err)
 	}
-	return unversionedObjToUpdate, nil
+
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(result.(map[string]interface{}), voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert merge result back: %v", err)
+	}
+	p.defaulter.Default(voutput)
+
+	uoutput, err := p.toUnversioned(voutput)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert to unversioned: %v", err)
+	}
+
+	return uoutput, nil
+}
+
+func (p *applyPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
```

Comment:
Will errors this returns get converted into the correct status error for human consumption?

Reply:
I'm not sure what you mean or what we should do. I'll look into it though.

Reply:
The error return stack only gives human friendly errors if the error implements a certain interface. So somewhere between here and there, something needs to convert the error into the correct object. Tracing back the call stack should answer the question.

Reply:
I've used the errors for debugging, here's what they look like:

```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "We managed to reach the end, congratulations!",
  "code": 500
}
```

Is that what you'd expect?